### ISSUE TITLE
Process crashed: Bad access due to protection failure

### ISSUE THREAD

**Issue by H0r53** [frida/frida#2039]
I am using frida to hook functions inside of a Shared Object that is used by an Android APK. I've calculated the addresses of functions within the Shared Object I am interested in and I have validated they are the correct addresses by dumping memory at those locations and matching the bytes with the shared object's assembly.

For some of my hooks everything works fine. However, for others I'll receive `Process crashed: Bad access due to protection failure`. For each address I'm interested in intercepting calls to, I set the hook as such.

```
    Interceptor.attach(ptr(addr), {
        onEnter(args)
        {
            console.log('Context information:');
            console.log('Context  : ' + JSON.stringify(this.context));
            console.log('Return   : ' + this.returnAddress);
            console.log('ThreadId : ' + this.threadId);
            console.log('Depth    : ' + this.depth);
            console.log('Errornr  : ' + this.err);    
        },
        onLeave(result)
        {
            console.log("Leaving")
        }
    })
```

In some cases, everything works perfectly fine. In other cases, the `console.log` statements print for the `onEnter` event but I receive `Process crashed: Bad access due to protection failure` before the `onLeave` event triggers. In yet other cases I receive the same bad access error before `onEnter` occurs.

Why is this happening? Can I adjust the protection of binary sections to avoid this? I'm not trying to access anything outside of the `.text` section of the binary as it is, so the protection error doesn't make sense.

All I want to do is dump the arguments (or really just do a hexdump of the stack) when a certain function is called or when the instruction pointer reaches a particular address.

**Comment by H0r53** [comment]
A follow up to this is, **can you setup frida to essentially "break" on a particular address and then do something like dump the current context**? My guess is frida is screwing up the address space of the process when creating a trampoline in certain places. I guess the converse question is **can you use a frida interceptor on any address, even if it isn't a function?**

In my case I'm very positive that I'm setting the Interceptor on a function address. However, the code is obfuscated so it may be using "untraditional" methods for adjusting the stack frame and etc.

**Comment by oleavr** [comment]
Using `Interceptor.attach(address, { onEnter() { ... }, onLeave() { ... } })` is very different from `Interceptor.attach(address, function () { ... })`. With the former you are asserting that `address` is first instruction of a function, whereas in the latter case you it can be any instruction. If you use the former syntax when you intend the latter, it will typically result in corruption of the location expected to contain the return address upon entry, which means `rsp` on x86_64, `lr` on arm64, etc. Perhaps this is the issue?

**Comment by H0r53** [comment]
Thanks for your input @oleavr ! I tried the `Interceptor.attach(address, function () { ... })` version and I'm receiving the same behavior as before. As soon as the address I'm attempting to intercept is encountered, I receive the **Process crashed: Bad access due to protection failure** error and the frida client crashes (frida server on the phone still runs).

Here are some more error details if they help.
```
Process crashed: Bad access due to protection failure

***
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'google/taimen/taimen:11/RP1A.201005.004.A1/6934943:user/release-keys'
Revision: 'rev_10'
ABI: 'arm64'
Timestamp: 2022-03-17 08:07:51-0600
pid: 28993, tid: 29118, name: RxCachedThreadS  >>> com.myapk <<<
uid: 10193
signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x7bcaa591fc
```


Is there an alternative approach to doing this? What I really want to do is basically dump the current context at a particular instruction / address and do some hexdumps of the stack and registers.

**Comment by H0r53** [comment]
@oleavr it's been several months and I'm still receiving this issue. I've even tried with a different phone and the latest frida server/client binaries. I don't have any issues when hooking some functions closer to the JNI exports, but when i want to either hook other functions or addresses that are a bit deeper in the call stack I continually receive an error similar to the following:

```
Process crashed: Bad access due to invalid address

***
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'google/bonito/bonito:9/PD2A.190115.029/5320088:user/release-keys'
Revision: 'MP1.0'
ABI: 'arm64'
pid: 8106, tid: 8259, name: RxCachedThreadS  >>> com.myapk <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x70a6c894f8
     x0  0000000000000000  x1  0000000000000000  x2  0000000000000000  x3  0000000000000003
     x4  0000000000000000  x5  2e6f747079ffffff  x6  0000000000008000  x7  0080000000000000
     x8  00000070a6c894f8  x9  0000006fa6e129c4  x10 0000000000000000  x11 0000000000000000
     x12 0000000000000000  x13 0000000000000008  x14 ffffffffffffffff  x15 000c4d3c611a8695
     x16 000000704d536cc0  x17 0000007050670b14  x18 0000006fa97ec42a  x19 0000000000000000
     x20 0000006fa97ed0b8  x21 0000006fa97ed108  x22 0000006fa97ed01c  x23 0000000000000001
     x24 0000006fa97f3588  x25 0000006fa97ed0a8  x26 0000006fc1a6d4a0  x27 0000000000000004
     x28 0000000000000008  x29 0000006fa97ed090
     sp  0000006fa97ebdc0  lr  0000006fa6c98f9c  pc  00000070a6c894f8

 backtrace:
     #00 pc 00000070a6c894f8  <unknown>
***
```

adb logcat reveals the following at the time of the crash.
```
5792   │ 07-13 12:49:00.211  8106  8259 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x70a6c894f8 in tid 8259 (RxCachedThre
       │ adS), pid 8106 (com.myapk)
5793   │ 07-13 12:49:00.219   799  1035 D CHRE    : @ 566.791: [ImuCal] Dynamic sensor configuration: high-performance.
5794   │ 07-13 12:49:00.221   799  8439 I sensors-hal: handle_indication_realtime:352,  SCHED_FIFO(10) for qmi_cbk
5795   │ 07-13 12:49:00.246  8449  8449 E DEBUG   : failed to readlink /proc/8259/fd/98: No such file or directory
5796   │ 07-13 12:49:00.284  8449  8449 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
5797   │ 07-13 12:49:00.285   897   897 I /system/bin/tombstoned: received crash request for pid 8259
5798   │ 07-13 12:49:00.286  8449  8449 I crash_dump64: performing dump of process 8106 (target tid = 8259)
5799   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
5800   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : Build fingerprint: 'google/bonito/bonito:9/PD2A.190115.029/5320088:user/release-keys'
5801   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : Revision: 'MP1.0'
5802   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : ABI: 'arm64'
5803   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : pid: 8106, tid: 8259, name: RxCachedThreadS  >>> com.myapk <<<
5804   │ 07-13 12:49:00.293  8449  8449 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x70a6c894f8
5805   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x0  0000000000000000  x1  0000000000000000  x2  0000000000000000  x3  0000000000000003
5806   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x4  0000000000000000  x5  2e6f747079ffffff  x6  0000000000008000  x7  0080000000000000
5807   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x8  00000070a6c894f8  x9  0000006fa6e129c4  x10 0000000000000000  x11 0000000000000000
5808   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x12 0000000000000000  x13 0000000000000008  x14 ffffffffffffffff  x15 000c4d3c611a8695
5809   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x16 000000704d536cc0  x17 0000007050670b14  x18 0000006fa97ec42a  x19 0000000000000000
5810   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x20 0000006fa97ed0b8  x21 0000006fa97ed108  x22 0000006fa97ed01c  x23 0000000000000001
5811   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x24 0000006fa97f3588  x25 0000006fa97ed0a8  x26 0000006fc1a6d4a0  x27 0000000000000004
5812   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     x28 0000000000000008  x29 0000006fa97ed090
5813   │ 07-13 12:49:00.293  8449  8449 F DEBUG   :     sp  0000006fa97ebdc0  lr  0000006fa6c98f9c  pc  00000070a6c894f8
5814   │ 07-13 12:49:00.294  8449  8449 F DEBUG   :
5815   │ 07-13 12:49:00.294  8449  8449 F DEBUG   : backtrace:
5816   │ 07-13 12:49:00.294  8449  8449 F DEBUG   :     #00 pc 00000070a6c894f8  <unknown>
```

Interestingly, the error is a `SIGSEGV` at address `0x70a6c894f8`, but the address I hooked was `0x6fa6dc9d4c`. So it looks like something else, presumably frida's trampoline, caused this error. I've tried both 

```
Interceptor.attach(ptr(addr), function() { })
```
and
```
Interceptor.attach(ptr(addr), {
        onEnter(args)
        {},
        onLeave(result)
        {}
})
```
The crash behavior is identical in both cases.

Note that this native library does use obfuscation techniques, so it isn't always maintaining a traditional stackframe. However, I don't have issues hooking functions/addresses closer to the JNI entry point routines. It's only after encountering what I imagine is anti-reversing code that this occurs.

Example: The backtrace provided only shows a single function call but there is absolutely a longer backtrace that isn't showing.

I thought the backtrace issue wouldn't matter if using interceptor to hook a specific address though. Why does the stackframe / backtrace matter if I just want to dump some memory when PC reaches a particular address?

Thank you for your time!

**Comment by H0r53** [comment]
Interestingly `SEGV_MAPERR` seems to indicate that the referenced PC isn't valid within the current address space (at least from my understanding). I checked `/proc/<pid>/maps` for this APK process and found that `0x70a6c894f8` is in fact not a valid address within currently mapped libraries. Where is this PC coming from? This value of course changes each time the app runs but the concept is still the same - for whatever reason whenever my frida hook is set around the location of functions I'm interested in, PC somehow appears to eventually become invalid.

**Comment by H0r53** [comment]
New note - I managed to use Frida Stalker instead of Interceptor and I've been able to dump memory when reaching the addresses I'm interested in without crashing the process. This is awesome and suits my needs, though it would appear that this issue is somehow related to Frida Interceptor. My guess is the binary is corrupting the stackframe by using non-traditional means, which Interceptor may be overwriting.

**Comment by Forterc** [comment]
I second this ^ - Frida-stalker with `iterator.putCallout(callbackFunc)` works alright, but interceptor fails in android. This happens in a specific use case - the app has anti reversing, the specific function prologue is messed up, there's stack corruption and polymoprhic assembly - but it's probably an issue with stalker in these cases, and not just a measure of that specific app.

@oleavr sorry for pinging buddy - chiming in to verify on my end this is a stalker issue. I'd suggest seeing if it's something you'd like to fix, and if so, I'm sure @H0r53 would be happy to demonstrate the specific case this happens for.

Thank you!

**Comment by mrupdate** [comment]
> New note - I managed to use Frida Stalker instead of Interceptor and I've been able to dump memory when reaching the addresses I'm interested in without crashing the process. This is awesome and suits my needs, though it would appear that this issue is somehow related to Frida Interceptor. My guess is the binary is corrupting the stackframe by using non-traditional means, which Interceptor may be overwriting.

@H0r53  Hello. Excuse me, do you use `Stalker.addCallProbe` instead of `Interceptor.attach`?

**Comment by H0r53** [comment]
@mrupdate I'm only having issues when attempting to use `Interceptor` in this scenario. Using `Stalker` has not presented any issues and has served as a reasonable workaround for now. Within Stalker I am not currently using `addCallProbe`, but instead I've instructed Stalker to follow a particular thread. Within Stalker's `transform` method I access register information using `iterator.putCallout()`.

**Comment by DarkSkull91** [comment]
Hi @H0r53, I'm facing the same issue but I'm a totally newbie with Frida. Can you explain in more detail how to face this situation? Or otherwise can you give me articles or documentation to handle this kind of problem?

Thanks

**Comment by H0r53** [comment]
@DarkSkull91 this issue could occur for several reasons so you're going to need to be more specific about what you've tried and what error is thrown, in which case you may as well open a new issue. My issue was related to frida interceptor crashing when trying to set hooks on certain addresses in an arm64 binary. The binary was hardened and using whitebox crypto, though I still feel like frida should have been and to intercept it.

**Comment by Irsathz** [comment]
@H0r53 
The same issues happened to me. Do you have any solutions?

**Comment by nicolaipre** [comment]
Try with `adb root` followed by starting the `frida-server`.
